### PR TITLE
Zilliz filter solution

### PR DESCRIPTION
**Pull request by hhy3**
expected results:
yfcc-10M

Our solution is based on vamana graph and tag partition. For each query, we handle it differently according to its tag's cardinality. We construct a vamana graph for every tag with cardinality larger than a threshold. For each query, we search with graph if its cardinality is large and search with brute force if its cardinality is small. Additionally, we organize queries tag by tag to enhance cache locality.

### CODE REVIEW

**Comment by liliu-z** [comment]
@harsha-simhadri Looks like people can continue contribute to this benchmark. Our indexing team is also interested in this and want to be part of it. Can you help to take a run? Appreciate it.

**Comment by ingberam** [comment]
> @harsha-simhadri Looks like people can continue contributing to this benchmark. Our indexing team is also interested in this and want to be part of it. Can you help to take a run? Appreciate it.

Hello, just as a heads up, we are planning to have a formal ongoing leaderboard, that will be updated periodically. The next update is planned to be at Feb. 1st, 2024. at this date, one of the organizers will run/merge all new algoritms, and publish the leaderboard.

**Comment by ingberam** [comment]
Regarding the current PR, please note that there are problems with the installation:

```
 python install.py --algorithm zilliz --neurips23track filter
Building base image...
Sending build context to Docker daemon  197.2MB
Step 1/9 : FROM ubuntu:jammy
 ---> 1f6ddc1b2547
Step 2/9 : RUN apt-get update && apt-get install -y python3-numpy python3-scipy python3-pip build-essential git axel wget
 ---> Using cache
 ---> 4bfa98b61265
Step 3/9 : RUN wget https://aka.ms/downloadazcopy-v10-linux && mv downloadazcopy-v10-linux azcopy.tgz && tar xzf azcopy.tgz --transform 's!^[^/]\+\($\|/\)!azcopy_folder\1!'
 ---> Using cache
 ---> 20fc3292ff8f
Step 4/9 : RUN cp azcopy_folder/azcopy /usr/bin
 ---> Using cache
 ---> 85343dfdeb1f
Step 5/9 : RUN pip3 install -U pip
 ---> Using cache
 ---> e2bf30e8a692
Step 6/9 : WORKDIR /home/app
 ---> Using cache
 ---> d9cdc86a4fc8
Step 7/9 : COPY requirements_py3.10.txt run_algorithm.py ./
 ---> Using cache
 ---> 6fa79adf9c1f
Step 8/9 : RUN pip3 install -r requirements_py3.10.txt
 ---> Using cache
 ---> 64344458270c
Step 9/9 : ENTRYPOINT ["python3", "-u", "run_algorithm.py"]
 ---> Using cache
 ---> 721df0ce4751
Successfully built 721df0ce4751
Successfully tagged neurips23:latest
Building algorithm images... with (1) processes
Building neurips23-filter-zilliz...
docker build  --rm -t neurips23-filter-zilliz -f neurips23/filter/zilliz/Dockerfile .
Sending build context to Docker daemon  197.2MB
Step 1/16 : FROM neurips23
 ---> 721df0ce4751
Step 2/16 : RUN apt update && apt install -y wget swig
 ---> Using cache
 ---> 625734611a51
Step 3/16 : RUN apt install -y git make cmake g++ libaio-dev libgoogle-perftools-dev libunwind-dev clang-format libboost-dev libboost-program-options-dev libmkl-full-dev libcpprest-dev python3.10
 ---> Running in 477d1e085001

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

Reading package lists...
Building dependency tree...
Reading state information...
g++ is already the newest version (4:11.2.0-1ubuntu1).
g++ set to manually installed.
libboost-dev is already the newest version (1.74.0.3ubuntu7).
libboost-dev set to manually installed.
make is already the newest version (4.3-4.1build1).
make set to manually installed.
The following additional packages will be installed:
  clang-format-14 cmake-data dh-elpa-helper emacsen-common icu-devtools
  intel-mkl-doc libaio1 libarchive13 libboost-atomic-dev
  libboost-atomic1.74-dev libboost-atomic1.74.0 libboost-chrono-dev
  libboost-chrono1.74-dev libboost-chrono1.74.0 libboost-date-time1.74-dev
  libboost-date-time1.74.0 libboost-filesystem-dev libboost-filesystem1.74-dev
  libboost-filesystem1.74.0 libboost-program-options1.74-dev
  libboost-program-options1.74.0 libboost-random-dev libboost-random1.74-dev
  libboost-random1.74.0 libboost-regex-dev libboost-regex1.74-dev
  libboost-regex1.74.0 libboost-serialization1.74-dev
  libboost-serialization1.74.0 libboost-system-dev libboost-system1.74-dev
  libboost-system1.74.0 libboost-thread-dev libboost-thread1.74-dev
  libboost-thread1.74.0 libbrotli-dev libclang-cpp14 libcpprest2.10 libcurl4
  libelf1 libgoogle-perftools4 libicu-dev libjsoncpp25 libllvm14 liblzma-dev
  libmkl-avx libmkl-avx2 libmkl-avx512 libmkl-avx512-mic
  libmkl-computational-dev libmkl-core libmkl-def libmkl-dev libmkl-gf-ilp64
  libmkl-gf-lp64 libmkl-gnu-thread libmkl-intel-ilp64 libmkl-intel-lp64
  libmkl-intel-thread libmkl-interface-dev libmkl-locale libmkl-mc libmkl-mc3
  libmkl-meta-computational libmkl-meta-interface libmkl-meta-threading
  libmkl-pgi-thread libmkl-rt libmkl-sequential libmkl-tbb-thread
  libmkl-threading-dev libmkl-vml-avx libmkl-vml-avx2 libmkl-vml-avx512
  libmkl-vml-avx512-mic libmkl-vml-cmpt libmkl-vml-def libmkl-vml-mc
  libmkl-vml-mc2 libmkl-vml-mc3 libomp-14-dev libomp-dev libomp5-14
  libpython3.10 libpython3.10-dev libpython3.10-minimal libpython3.10-stdlib
  librhash0 libssl-dev libssl3 libtbb2 libtbb2-dev libtbbmalloc2
  libtcmalloc-minimal4 libunwind8 libuv1 libwebsocketpp-dev pkg-config
  python3.10-dev python3.10-minimal
Suggested packages:
  cmake-doc ninja-build cmake-format gettext-base git-daemon-run
  | git-daemon-sysvinit git-doc git-email git-gui gitk gitweb git-cvs
  git-mediawiki git-svn lrzip icu-doc liblzma-doc libomp-14-doc libssl-doc
  libtbb2-doc python3.10-venv python3.10-doc binfmt-support
Recommended packages:
  libmkl-rt:i386 libmkl-interface-dev:i386 libmkl-threading-dev:i386
  libmkl-computational-dev:i386
The following NEW packages will be installed:
  clang-format clang-format-14 cmake cmake-data dh-elpa-helper emacsen-common
  icu-devtools intel-mkl-doc libaio-dev libaio1 libarchive13
  libboost-atomic-dev libboost-atomic1.74-dev libboost-atomic1.74.0
  libboost-chrono-dev libboost-chrono1.74-dev libboost-chrono1.74.0
  libboost-date-time1.74-dev libboost-date-time1.74.0 libboost-filesystem-dev
  libboost-filesystem1.74-dev libboost-filesystem1.74.0
  libboost-program-options-dev libboost-program-options1.74-dev
  libboost-program-options1.74.0 libboost-random-dev libboost-random1.74-dev
  libboost-random1.74.0 libboost-regex-dev libboost-regex1.74-dev
  libboost-regex1.74.0 libboost-serialization1.74-dev
  libboost-serialization1.74.0 libboost-system-dev libboost-system1.74-dev
  libboost-system1.74.0 libboost-thread-dev libboost-thread1.74-dev
  libboost-thread1.74.0 libbrotli-dev libclang-cpp14 libcpprest-dev
  libcpprest2.10 libcurl4 libelf1 libgoogle-perftools-dev libgoogle-perftools4
  libicu-dev libjsoncpp25 libllvm14 liblzma-dev libmkl-avx libmkl-avx2
  libmkl-avx512 libmkl-avx512-mic libmkl-computational-dev libmkl-core
  libmkl-def libmkl-dev libmkl-full-dev libmkl-gf-ilp64 libmkl-gf-lp64
  libmkl-gnu-thread libmkl-intel-ilp64 libmkl-intel-lp64 libmkl-intel-thread
  libmkl-interface-dev libmkl-locale libmkl-mc libmkl-mc3
  libmkl-meta-computational libmkl-meta-interface libmkl-meta-threading
  libmkl-pgi-thread libmkl-rt libmkl-sequential libmkl-tbb-thread
  libmkl-threading-dev libmkl-vml-avx libmkl-vml-avx2 libmkl-vml-avx512
  libmkl-vml-avx512-mic libmkl-vml-cmpt libmkl-vml-def libmkl-vml-mc
  libmkl-vml-mc2 libmkl-vml-mc3 libomp-14-dev libomp-dev libomp5-14 librhash0
  libssl-dev libtbb2 libtbb2-dev libtbbmalloc2 libtcmalloc-minimal4
  libunwind-dev libunwind8 libuv1 libwebsocketpp-dev pkg-config
The following packages will be upgraded:
  git libpython3.10 libpython3.10-dev libpython3.10-minimal
  libpython3.10-stdlib libssl3 python3.10 python3.10-dev python3.10-minimal
9 upgraded, 101 newly installed, 0 to remove and 72 not upgraded.
Need to get 405 MB of archives.
After this operation, 2273 MB of additional disk space will be used.
Ign:1 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 python3.10-dev amd64 3.10.12-1~22.04.2
Ign:2 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 libpython3.10-dev amd64 3.10.12-1~22.04.2
Ign:3 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 libpython3.10 amd64 3.10.12-1~22.04.2
Get:4 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 libssl3 amd64 3.0.2-0ubuntu1.12 [1901 kB]
Err:1 http://security.ubuntu.com/ubuntu jammy-updates/main amd64 python3.10-dev amd64 3.10.12-1~22.04.2
  404  Not Found [IP: 91.189.91.83 80]
Err:2 http://security.ubuntu.com/ubuntu jammy-updates/main amd64 libpython3.10-dev amd64 3.10.12-1~22.04.2
  404  Not Found [IP: 91.189.91.83 80]
Err:3 http://security.ubuntu.com/ubuntu jammy-updates/main amd64 libpython3.10 amd64 3.10.12-1~22.04.2
  404  Not Found [IP: 91.189.91.83 80]
Ign:5 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 python3.10 amd64 3.10.12-1~22.04.2
Ign:6 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 libpython3.10-stdlib amd64 3.10.12-1~22.04.2
Ign:7 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 python3.10-minimal amd64 3.10.12-1~22.04.2
Ign:8 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 libpython3.10-minimal amd64 3.10.12-1~22.04.2
Get:9 http://archive.ubuntu.com/ubuntu jammy/main amd64 libelf1 amd64 0.186-1build1 [51.0 kB]
Get:10 http://archive.ubuntu.com/ubuntu jammy/main amd64 libuv1 amd64 1.43.0-1 [93.1 kB]
Get:11 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 libllvm14 amd64 1:14.0.0-1ubuntu1.1 [24.0 MB]
Err:5 http://security.ubuntu.com/ubuntu jammy-updates/main amd64 python3.10 amd64 3.10.12-1~22.04.2
  404  Not Found [IP: 91.189.91.83 80]
Err:6 http://security.ubuntu.com/ubuntu jammy-updates/main amd64 libpython3.10-stdlib amd64 3.10.12-1~22.04.2
  404  Not Found [IP: 91.189.91.83 80]
Err:7 http://security.ubuntu.com/ubuntu jammy-updates/main amd64 python3.10-minimal amd64 3.10.12-1~22.04.2
  404  Not Found [IP: 91.189.91.83 80]
Err:8 http://security.ubuntu.com/ubuntu jammy-updates/main amd64 libpython3.10-minimal amd64 3.10.12-1~22.04.2
  404  Not Found [IP: 91.189.91.83 80]
Get:12 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 libclang-cpp14 amd64 1:14.0.0-1ubuntu1.1 [12.1 MB]
Get:13 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 clang-format-14 amd64 1:14.0.0-1ubuntu1.1 [50.8 kB]
Get:14 http://archive.ubuntu.com/ubuntu jammy/universe amd64 clang-format amd64 1:14.0-55~exp2 [3694 B]
Get:15 http://archive.ubuntu.com/ubuntu jammy/main amd64 libarchive13 amd64 3.6.0-1ubuntu1 [368 kB]
Ign:16 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 libcurl4 amd64 7.81.0-1ubuntu1.14
Get:17 http://archive.ubuntu.com/ubuntu jammy/main amd64 libjsoncpp25 amd64 1.9.5-3 [80.0 kB]
Get:18 http://archive.ubuntu.com/ubuntu jammy/main amd64 librhash0 amd64 1.4.2-1ubuntu1 [125 kB]
Get:19 http://archive.ubuntu.com/ubuntu jammy/main amd64 dh-elpa-helper all 2.0.9ubuntu1 [7610 B]
Get:20 http://archive.ubuntu.com/ubuntu jammy/main amd64 emacsen-common all 3.0.4 [14.9 kB]
Get:21 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 cmake-data all 3.22.1-1ubuntu1.22.04.1 [1913 kB]
Get:22 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 cmake amd64 3.22.1-1ubuntu1.22.04.1 [5013 kB]
Err:16 http://security.ubuntu.com/ubuntu jammy-updates/main amd64 libcurl4 amd64 7.81.0-1ubuntu1.14
  404  Not Found [IP: 91.189.91.83 80]
Get:23 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 git amd64 1:2.34.1-1ubuntu1.10 [3166 kB]
Get:24 http://archive.ubuntu.com/ubuntu jammy/main amd64 icu-devtools amd64 70.1-2 [197 kB]
Get:25 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 intel-mkl-doc all 2020.4.304-2ubuntu3 [3233 kB]
Get:26 http://archive.ubuntu.com/ubuntu jammy/main amd64 libaio1 amd64 0.3.112-13build1 [7176 B]
Get:27 http://archive.ubuntu.com/ubuntu jammy/main amd64 libaio-dev amd64 0.3.112-13build1 [21.2 kB]
Get:28 http://archive.ubuntu.com/ubuntu jammy/main amd64 libboost-atomic1.74.0 amd64 1.74.0-14ubuntu3 [224 kB]
Get:29 http://archive.ubuntu.com/ubuntu jammy/main amd64 libboost-atomic1.74-dev amd64 1.74.0-14ubuntu3 [221 kB]
Get:30 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libboost-atomic-dev amd64 1.74.0.3ubuntu7 [3544 B]
Get:31 http://archive.ubuntu.com/ubuntu jammy/main amd64 libboost-chrono1.74.0 amd64 1.74.0-14ubuntu3 [232 kB]
Get:32 http://archive.ubuntu.com/ubuntu jammy/main amd64 libboost-chrono1.74-dev amd64 1.74.0-14ubuntu3 [239 kB]
Get:33 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libboost-chrono-dev amd64 1.74.0.3ubuntu7 [3854 B]
Get:34 http://archive.ubuntu.com/ubuntu jammy/main amd64 libboost-date-time1.74.0 amd64 1.74.0-14ubuntu3 [221 kB]
Get:35 http://archive.ubuntu.com/ubuntu jammy/main amd64 libboost-serialization1.74.0 amd64 1.74.0-14ubuntu3 [327 kB]
Get:36 http://archive.ubuntu.com/ubuntu jammy/main amd64 libboost-serialization1.74-dev amd64 1.74.0-14ubuntu3 [375 kB]
Get:37 http://archive.ubuntu.com/ubuntu jammy/main amd64 libboost-date-time1.74-dev amd64 1.74.0-14ubuntu3 [226 kB]
Get:38 http://archive.ubuntu.com/ubuntu jammy/main amd64 libboost-filesystem1.74.0 amd64 1.74.0-14ubuntu3 [264 kB]
Get:39 http://archive.ubuntu.com/ubuntu jammy/main amd64 libboost-system1.74.0 amd64 1.74.0-14ubuntu3 [221 kB]
Get:40 http://archive.ubuntu.com/ubuntu jammy/main amd64 libboost-system1.74-dev amd64 1.74.0-14ubuntu3 [218 kB]
Get:41 http://archive.ubuntu.com/ubuntu jammy/main amd64 libboost-filesystem1.74-dev amd64 1.74.0-14ubuntu3 [287 kB]
Get:42 http://archive.ubuntu.com/ubuntu jammy/main amd64 libboost-filesystem-dev amd64 1.74.0.3ubuntu7 [3280 B]
Get:43 http://archive.ubuntu.com/ubuntu jammy/main amd64 libboost-program-options1.74.0 amd64 1.74.0-14ubuntu3 [311 kB]
Get:44 http://archive.ubuntu.com/ubuntu jammy/main amd64 libboost-program-options1.74-dev amd64 1.74.0-14ubuntu3 [380 kB]
Get:45 http://archive.ubuntu.com/ubuntu jammy/main amd64 libboost-program-options-dev amd64 1.74.0.3ubuntu7 [3266 B]
Get:46 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libboost-random1.74.0 amd64 1.74.0-14ubuntu3 [231 kB]
Get:47 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libboost-random1.74-dev amd64 1.74.0-14ubuntu3 [231 kB]
Get:48 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libboost-random-dev amd64 1.74.0.3ubuntu7 [3258 B]
Get:49 http://archive.ubuntu.com/ubuntu jammy/main amd64 libboost-regex1.74.0 amd64 1.74.0-14ubuntu3 [511 kB]
Get:50 http://archive.ubuntu.com/ubuntu jammy/main amd64 libicu-dev amd64 70.1-2 [11.6 MB]
Get:51 http://archive.ubuntu.com/ubuntu jammy/main amd64 libboost-regex1.74-dev amd64 1.74.0-14ubuntu3 [596 kB]
Get:52 http://archive.ubuntu.com/ubuntu jammy/main amd64 libboost-regex-dev amd64 1.74.0.3ubuntu7 [3510 B]
Get:53 http://archive.ubuntu.com/ubuntu jammy/main amd64 libboost-system-dev amd64 1.74.0.3ubuntu7 [3390 B]
Get:54 http://archive.ubuntu.com/ubuntu jammy/main amd64 libboost-thread1.74.0 amd64 1.74.0-14ubuntu3 [262 kB]
Get:55 http://archive.ubuntu.com/ubuntu jammy/main amd64 libboost-thread1.74-dev amd64 1.74.0-14ubuntu3 [272 kB]
Get:56 http://archive.ubuntu.com/ubuntu jammy/main amd64 libboost-thread-dev amd64 1.74.0.3ubuntu7 [3266 B]
Get:57 http://archive.ubuntu.com/ubuntu jammy/main amd64 libbrotli-dev amd64 1.0.9-2build6 [337 kB]
Get:58 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libcpprest2.10 amd64 2.10.18-1build2 [841 kB]
Get:59 http://archive.ubuntu.com/ubuntu jammy/main amd64 libtcmalloc-minimal4 amd64 2.9.1-0ubuntu3 [98.2 kB]
Get:60 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 libunwind8 amd64 1.3.2-2build2.1 [54.5 kB]
Get:61 http://archive.ubuntu.com/ubuntu jammy/main amd64 libgoogle-perftools4 amd64 2.9.1-0ubuntu3 [212 kB]
Get:62 http://archive.ubuntu.com/ubuntu jammy/main amd64 liblzma-dev amd64 5.2.5-2ubuntu1 [159 kB]
Get:63 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 libunwind-dev amd64 1.3.2-2build2.1 [1883 kB]
Get:64 http://archive.ubuntu.com/ubuntu jammy/main amd64 libgoogle-perftools-dev amd64 2.9.1-0ubuntu3 [470 kB]
Get:65 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-locale amd64 2020.4.304-2ubuntu3 [30.7 kB]
Get:66 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-avx amd64 2020.4.304-2ubuntu3 [13.6 MB]
Get:67 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-avx2 amd64 2020.4.304-2ubuntu3 [11.9 MB]
Get:68 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-avx512 amd64 2020.4.304-2ubuntu3 [13.2 MB]
Get:69 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-avx512-mic amd64 2020.4.304-2ubuntu3 [14.0 MB]
Get:70 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-def amd64 2020.4.304-2ubuntu3 [9921 kB]
Get:71 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-vml-def amd64 2020.4.304-2ubuntu3 [2470 kB]
Get:72 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-core amd64 2020.4.304-2ubuntu3 [15.8 MB]
Get:73 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-mc amd64 2020.4.304-2ubuntu3 [11.6 MB]
Get:74 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-mc3 amd64 2020.4.304-2ubuntu3 [11.9 MB]
Get:75 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-vml-avx amd64 2020.4.304-2ubuntu3 [4226 kB]
Get:76 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-vml-avx2 amd64 2020.4.304-2ubuntu3 [3805 kB]
Get:77 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-vml-avx512 amd64 2020.4.304-2ubuntu3 [3639 kB]
Get:78 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-vml-avx512-mic amd64 2020.4.304-2ubuntu3 [3905 kB]
Get:79 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-vml-cmpt amd64 2020.4.304-2ubuntu3 [2213 kB]
Get:80 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-vml-mc amd64 2020.4.304-2ubuntu3 [3753 kB]
Get:81 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-vml-mc2 amd64 2020.4.304-2ubuntu3 [3751 kB]
Get:82 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-vml-mc3 amd64 2020.4.304-2ubuntu3 [3755 kB]
Get:83 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-meta-computational amd64 2020.4.304-2ubuntu3 [7464 B]
Get:84 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-computational-dev amd64 2020.4.304-2ubuntu3 [110 MB]
Get:85 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-intel-lp64 amd64 2020.4.304-2ubuntu3 [2006 kB]
Get:86 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-intel-ilp64 amd64 2020.4.304-2ubuntu3 [1860 kB]
Get:87 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-gf-lp64 amd64 2020.4.304-2ubuntu3 [2018 kB]
Get:88 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-gf-ilp64 amd64 2020.4.304-2ubuntu3 [1861 kB]
Get:89 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-meta-interface amd64 2020.4.304-2ubuntu3 [7404 B]
Get:90 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-intel-thread amd64 2020.4.304-2ubuntu3 [9053 kB]
Get:91 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-gnu-thread amd64 2020.4.304-2ubuntu3 [4841 kB]
Get:92 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-sequential amd64 2020.4.304-2ubuntu3 [4989 kB]
Get:93 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-tbb-thread amd64 2020.4.304-2ubuntu3 [5880 kB]
Get:94 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-pgi-thread amd64 2020.4.304-2ubuntu3 [5410 kB]
Get:95 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-meta-threading amd64 2020.4.304-2ubuntu3 [7408 B]
Get:96 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-rt amd64 2020.4.304-2ubuntu3 [1214 kB]
Get:97 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-interface-dev amd64 2020.4.304-2ubuntu3 [5109 kB]
Get:98 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-threading-dev amd64 2020.4.304-2ubuntu3 [25.8 MB]
Get:99 http://archive.ubuntu.com/ubuntu jammy/main amd64 pkg-config amd64 0.29.2-1ubuntu3 [48.2 kB]
Get:100 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-dev amd64 2020.4.304-2ubuntu3 [561 kB]
Get:101 http://archive.ubuntu.com/ubuntu jammy/multiverse amd64 libmkl-full-dev amd64 2020.4.304-2ubuntu3 [18.5 kB]
Get:102 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 libomp5-14 amd64 1:14.0.0-1ubuntu1.1 [389 kB]
Get:103 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 libomp-14-dev amd64 1:14.0.0-1ubuntu1.1 [347 kB]
Get:104 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 libssl-dev amd64 3.0.2-0ubuntu1.12 [2373 kB]
Get:105 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libtbbmalloc2 amd64 2021.5.0-7ubuntu2 [49.6 kB]
Get:106 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libtbb2 amd64 2020.3-1ubuntu3 [101 kB]
Get:107 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libtbb2-dev amd64 2020.3-1ubuntu3 [278 kB]
Get:108 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libwebsocketpp-dev amd64 0.8.2-4 [119 kB]
Get:109 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libcpprest-dev amd64 2.10.18-1build2 [158 kB]
Get:110 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libomp-dev amd64 1:14.0-55~exp2 [3074 B]
Fetched 392 MB in 8s (51.1 MB/s)
E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/p/python3.10/python3.10-dev_3.10.12-1%7e22.04.2_amd64.deb  404  Not Found [IP: 91.189.91.83 80]
E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/p/python3.10/libpython3.10-dev_3.10.12-1%7e22.04.2_amd64.deb  404  Not Found [IP: 91.189.91.83 80]
E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/p/python3.10/libpython3.10_3.10.12-1%7e22.04.2_amd64.deb  404  Not Found [IP: 91.189.91.83 80]
E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/p/python3.10/python3.10_3.10.12-1%7e22.04.2_amd64.deb  404  Not Found [IP: 91.189.91.83 80]
E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/p/python3.10/libpython3.10-stdlib_3.10.12-1%7e22.04.2_amd64.deb  404  Not Found [IP: 91.189.91.83 80]
E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/p/python3.10/python3.10-minimal_3.10.12-1%7e22.04.2_amd64.deb  404  Not Found [IP: 91.189.91.83 80]
E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/p/python3.10/libpython3.10-minimal_3.10.12-1%7e22.04.2_amd64.deb  404  Not Found [IP: 91.189.91.83 80]
E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/c/curl/libcurl4_7.81.0-1ubuntu1.14_amd64.deb  404  Not Found [IP: 91.189.91.83 80]
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
The command '/bin/sh -c apt install -y git make cmake g++ libaio-dev libgoogle-perftools-dev libunwind-dev clang-format libboost-dev libboost-program-options-dev libmkl-full-dev libcpprest-dev python3.10' returned a non-zero code: 100


Install Status:
{'neurips23-filter-zilliz': 'fail'}

```

**Comment by hhy3** [comment]
@ingberam updated, can you run it again?

**Comment by ingberam** [comment]
@hhy3 suggestion: it seems that you only use `bow_id_selector` at index build time (line 127) in order to intersect the inverted lists. Why not simplify that and compute it in python, with zero effect on search QPS? This way you can avoid building that additional binary, and the Dockerfile can have only 5 lines like [this](https://github.com/harsha-simhadri/big-ann-benchmarks/blob/main/neurips23/filter/pinecone/Dockerfile), instead of 18 lines (with complexities such as the one above)?

**Comment by hhy3** [comment]
@ingberam Thanks for your suggestion. Codes are updated.

**Comment by ingberam** [comment]
Code installs smoothly with the simplified Dockerfile.
But the index build keeps failing due to insufficient disk space. How much disk space does it need??

**Comment by hhy3** [comment]
> Code installs smoothly with the simplified Dockerfile. But the index build keeps failing due to insufficient disk space. How much disk space does it need??

Seems like diskann saved some temporal data on disk. I'll fix it soon

**Comment by ingberam** [comment]
I was able to clear some space on the VM so the index runs. The index build process generates 58GB of data, comprising of 6,733 folders, each representing a diskann index I assume. It seems that in each folder is a separate diskann index. The folder seems to contain the raw vectors twice, causing the large space requirements. It would be best to avoid redundant saves.

The main block right now is this - after these indexes are built, the main index does not load into memory and docker crashes with OOM (error 137). 
Note that on the VM, there are some processes running that restrict the memory for the docker process. This restriction is more pronounced on the standard Azure VM that is used for evaluation (compared to e.g. GCP). Specifically, the memory available for docker is 14.3-14.5GB.

Bottom line - the index does not currently run. Try to reduce the memory consumption.

**Comment by hhy3** [comment]
@ingberam I've fixed the disk issue and modified some params to limit memory usage under 13G so it should work. Can you run it again?

**Comment by ingberam** [comment]
```
zilliz,filter_searcher_R12_L100_ef16,yfcc-10M,10,85456.02685170338,0.0,3336.493741750717,12929620.0,0,0,filter,0.8922549999999999
zilliz,filter_searcher_R12_L100_ef18,yfcc-10M,10,84839.95255456145,0.0,3336.493741750717,12929620.0,0,0,filter,0.899427
zilliz,filter_searcher_R12_L100_ef19,yfcc-10M,10,84596.41921284643,0.0,3336.493741750717,12929620.0,0,0,filter,0.902448
zilliz,filter_searcher_R12_L100_ef20,yfcc-10M,10,83520.42756810682,0.0,3336.493741750717,12929620.0,0,0,filter,0.905295
zilliz,filter_searcher_R12_L100_ef22,yfcc-10M,10,83199.00596890757,0.0,3336.493741750717,12929620.0,0,0,filter,0.910399
zilliz,filter_searcher_R12_L100_ef24,yfcc-10M,10,82993.20589371308,0.0,3336.493741750717,12929620.0,0,0,filter,0.9147959999999999
zilliz,filter_searcher_R12_L100_ef26,yfcc-10M,10,82555.9874635793,0.0,3336.493741750717,12929620.0,0,0,filter,0.918323
zilliz,filter_searcher_R12_L100_ef28,yfcc-10M,10,82157.64741371037,0.0,3336.493741750717,12929620.0,0,0,filter,0.921776
zilliz,filter_searcher_R12_L100_ef32,yfcc-10M,10,81347.09086920843,0.0,3336.493741750717,12929620.0,0,0,filter,0.927407
zilliz,filter_searcher_R12_L100_ef36,yfcc-10M,10,80531.72430040495,0.0,3336.493741750717,12929620.0,0,0,filter,0.9319660000000001
```

(also, pls fix the conflicts so that we can merge)

**Comment by hhy3** [comment]
@ingberam Fixed, thanks.
